plugins: fix compilation if FLB_SQLDB (sqlite3) is disabled

ThomasDevoogdt opened the pull request:
Fixes:
- https://github.com/fluent/fluent-bit/issues/9757.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - SQLDB support is now optional; enabling it restores partial-match checklist behavior and improved blob upload tracking.
  - Added a new post-upload action: "Add Suffix".

- **Bug Fixes**
  - Non-database builds now fail fast for DB-only modes/events to avoid unexpected behavior.
  - Improved initialization and cleanup to prevent leftover state on failures.

- **Tests**
  - CI and unit tests now run with DB-enabled and DB-disabled variants; DB-only tests are conditional.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

patrick-stephens commented:
Let's make sure this compiles for all existing targets as well.

Would it be better to disable the features within the plugins that need the DB rather than the whole plugin @leonardo-albertovich ? It feels a bit like a large hammer, e.g. no tail input even if you don't use db. We would have to make the config options and any other usage conditional though so it may be worse.

ThomasDevoogdt commented:
> Let's make sure this compiles for all existing targets as well.

How can it not? Currently, FLB_SQLDB is enabled by default, and will stay like that. The problem is the other way around, if not enabled, then compilation breaks. 

> Would it be better to disable the features within the plugins that need the DB rather than the whole plugin @leonardo-albertovich ? It feels a bit like a large hammer, e.g. no tail input even if you don't use db. We would have to make the config options and any other usage conditional though so it may be worse.

For me fine, but I would do that on a per feature basis. Perhaps just continue with this PR, and then fine-tune some features that might compile with some small fixups.

In general, it would be much better if all options are toggled in the automated tests, on one reference compilation system. Perhaps even by incremental builds. But either way, I just want to fix compilation now.

ThomasDevoogdt committed:
in_blob: fix compilation if FLB_SQLDB (sqlite3) is disabled

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt committed:
filter_checklist: fix compilation if FLB_SQLDB (sqlite3) is disabled

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt committed:
out_azure_blob: fix compilation if FLB_SQLDB (sqlite3) is disabled

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt commented:
@patrick-stephens @edsiper I changed this PR so that plugins are still compiled, but without database support. I hope this answers https://github.com/fluent/fluent-bit/pull/10239#issuecomment-2826839994. 

Tested by doing this:
```
cmake -GNinja -DFLB_PREFER_SYSTEM_LIBS=ON -DFLB_SQLDB=OFF ../
```

ThomasDevoogdt commented:
> Not sure if the CI failure is relevant or something else

I just added `#ifdef FLB_HAVE_SQLDB`, which is true by default, so I don't think it's relevant.

ThomasDevoogdt commented:
@edsiper All merge checks seems to be passing. Can you consider this PR? This will also close #9757.

leonardo-albertovich commented on include/fluent-bit/flb_event.h:
```diff
@@ -29,7 +29,10 @@
 #define FLB_EVENT_TYPE_TRACES     FLB_INPUT_CHUNK_TYPE_TRACES
 #define FLB_EVENT_TYPE_PROFILES   FLB_INPUT_CHUNK_TYPE_PROFILES
+
+#ifdef FLB_HAVE_SQLDB
```
Please remove this conditional, this should be handled at a different level.

ThomasDevoogdt replied:
I will restore it, can you answer on https://github.com/fluent/fluent-bit/pull/10239#issuecomment-2906988713, not clear where I should add a warning to indicate that this is not supported.

ThomasDevoogdt replied:
I've restored it.

leonardo-albertovich reviewed (changes requested):
In the case of `FLB_EVENT_TYPE_BLOBS` this constant should not be removed if don't have database support.

In the case of `filter_checklist` we should issue an error if `mode` is set to `partial` and `FLB_HAVE_SQLDB` is not defined, we cannot accept the patch as is because it would allow fluent-bit to build but cause a silent failure which would be way worse.

In the case of `in_blob` since it cannot operate without a database it should be disabled in the build script.

ThomasDevoogdt commented:
> In the case of `FLB_EVENT_TYPE_BLOBS` this constant should not be removed if don't have database support.

I can do that, but where exactly should I warn about it? 
Should I add one in `flb_input_blob_file_register`?

> In the case of `filter_checklist` we should issue an error if `mode` is set to `partial` and `FLB_HAVE_SQLDB` is not defined, we cannot accept the patch as is because it would allow fluent-bit to build but cause a silent failure which would be way worse.

Is it fine to add a warning in `init_config` in `plugins/filter_checklist/checklist.c`.

> In the case of `in_blob` since it cannot operate without a database it should be disabled in the build script.

But why do I see then `#ifdef FLB_HAVE_SQLDB` checks in the in_blob plugin if it should be disabled anyway? My commit just fixed one missing `#ifdef FLB_HAVE_SQLDB` which was forgotten.

See: https://github.com/search?q=repo%3Afluent%2Ffluent-bit%20path%3A%2F%5Eplugins%5C%2Fin_blob%5C%2F%2F%20FLB_HAVE_SQLDB&type=code.

patrick-stephens commented:
I'd probably be interested in an in-memory only option as well to remove more dependencies for hardening in those places where we need that and just ephemeral usage.

ThomasDevoogdt commented:
@leonardo-albertovich some of your remarks were fixed, can you have a look again?

ThomasDevoogdt committed:
blob_db: fix missing methods if not FLB_HAVE_SQLDB

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt commented:
@leonardo-albertovich I've updated this PR. Some new errors were fixed since. Can you have a look and eventually merge this?

ThomasDevoogdt committed:
workflows: build with FLB_SQLDB=On|Off

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt committed:
tests: runtime: filter_checklist: fix flb_test_mode_partial if !FLB_HAVE_SQLDB

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

ThomasDevoogdt committed:
tests: runtime_shell: exclude in_tail_expect if not FLB_HAVE_SQLDB

Signed-off-by: Thomas Devoogdt <thomas@devoogdt.com>

cosmo0920 reviewed (commented):
This PR seems good but we're not able to add :+1: because of failing unit testing.

edsiper commented:
I think in_blob, flb_blob_db must be disabled if sqlite is not available, since they must have the state.

Out S3 and out_azure will need to be aware of the lack of flb_blob_db so they can disable blob handling support

ThomasDevoogdt commented:
@edsiper 

> I think in_blob, flb_blob_db must be disabled if sqlite is not available, since they must have the state.

But why do I then see a bunch of other FLB_HAVE_SQLDB declaratives in flb_blob_db? 

e.g.
https://github.com/fluent/fluent-bit/blob/ed06c31991fb0c5e3f59677ceb63cabbf50ed1a7/plugins/in_blob/blob.c#L641

If it has to be disabled, do you expect me to drop all existing FLB_HAVE_SQLDB declaratives? Because I somehow expect that the original code should just compile, but was somehow broken over time. What I try to fix now.

leonardo-albertovich commented:
`in_blob` doesn't need to be disabled because it can operate without a database
`flb_blob_db` does't need to be excluded because when `FLB_HAVE_SQLDB` isn't detected it impleemnts the exact same functions except they all return `FLB_BLOB_DB_ERROR_NO_BACKEND_AVAILABLE` as the error code
`out_azure_blob` doesn't need to be disabled because when `FLB_HAVE_SQLDB` isn't detected it just doesn't process blobs
`out_s3` doesn't need to be disabled because when `FLB_HAVE_SQLDB` isn't detected it just doesn't process blobs

That's of course as far as I know and as a side note, the test case issue is unrelated to this feature because it's `flb-rt-core_chunk_trace`.

leonardo-albertovich reviewed (changes requested):
There's a minor nitpick but overall it's good to go.

I think replicating the `flb_blob_db` mechanism where the functions are defined but all of them return a specific error in `azure_blob_db.c`  would make more sense than adding most of the pre-processor conditionals in the callers but TBH that component as a whole should be replaced by `flb_blob_db` which is the reason why the functionality was extracted instead of implemented in `out_s3` originally so I don't think it'd be reasonable to get too pedantic here.

edsiper commented:
moving this until the next milestone, some parts are still moving

cosmo0920 reviewed (commented):
This looks good to me.
However, this review comment seems not to be addressed:
https://github.com/fluent/fluent-bit/pull/10239/files#r2492077239
Is it enough with this implementation?

ThomasDevoogdt commented:
> We fixed no left device error on our CI. So, could you rebase off current master? We need to confirm all of the CI tasks are passed before processing review.

Rebased, and all tests seems to be fine. Can we merge?

leonardo-albertovich commented:
I'll review this PR again tomorrow. @ThomasDevoogdt if you have any questions that were unanswered (I think I missed some) please repeat them so I can address them tomorrow.

ThomasDevoogdt commented:
@patrick-stephens @edsiper Again a week later. Can this be merged please?

ThomasDevoogdt commented:
@celalettin1286 @cosmo0920 @edsiper @niedbalski @patrick-stephens ping again ...


How many times do I still have to rebase this until someone merges this PR? This PR is a year old! I'm starting to loose my patience here.

ThomasDevoogdt commented:
@leonardo-albertovich ping

ThomasDevoogdt commented:
@leonardo-albertovich ping

patrick-stephens commented:
This is fine by me so will ping @cosmo0920 , I don't think @leonardo-albertovich has time for this at the moment

ThomasDevoogdt commented:
@patrick-stephens oké

ThomasDevoogdt commented:
@leonardo-albertovich @patrick-stephens ping

I would appreciate it if you guys just review it and merge it.

patrick-stephens commented:
I'm afraid I am unable to merge without @edsiper or @cosmo0920 review as codeowners
